src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -21,133 +21,505 @@
             }
-        };
-        reset = new Runnable()
+
+            if (dynamicSampleUpdateInterval != newDynamicSampleUpdateInterval || updateSamplesScheduler == null)
+            {
+                cancelAndWait(updateSamplesScheduler, Math.max(1, dynamicSampleUpdateInterval), TimeUnit.MILLISECONDS);
+                if (newDynamicSampleUpdateInterval > 0)
+                    updateSamplesScheduler = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(this::updateSamples, newDynamicSampleUpdateInterval, newDynamicSampleUpdateInterval, TimeUnit.MILLISECONDS);
+            }
+        }
+
+        dynamicUpdateInterval = newDynamicUpdateInternal;
+        dynamicSampleUpdateInterval = newDynamicSampleUpdateInterval;
+        dynamicBadnessThreshold = newDynamicBadnessThreshold;
+
+        if (dynamicSampleUpdateInterval > 0)
+            probeRateLimiter.setRate(dynamicSampleUpdateInterval);
+    }
+
+    public synchronized void open()
+    {
+        applyConfigChanges(DatabaseDescriptor.getDynamicUpdateInterval(),
+                           DatabaseDescriptor.getDynamicSampleUpdateInterval(),
+                           DatabaseDescriptor.getDynamicBadnessThreshold());
+
+        MBeanWrapper.instance.registerMBean(this, mbeanName);
+        mbeanRegistered = true;
+    }
+
+    public synchronized void close()
+    {
+        cancelAndWait(updateScoresScheduler, Math.max(1, dynamicUpdateInterval), TimeUnit.MILLISECONDS);
+        cancelAndWait(updateSamplesScheduler, Math.max(1, dynamicSampleUpdateInterval), TimeUnit.MILLISECONDS);
+        updateScoresScheduler = null;
+        updateSamplesScheduler = null;
+
+        for (AnnotatedMeasurement measurement : samples.values())
         {
-            public void run()
+            cancelAndWait(measurement.probeFuture, PING.getTimeout(), TimeUnit.MILLISECONDS);
+
+            measurement.millisSinceLastMeasure.set(0);
+            measurement.millisSinceLastRequest.set(MAX_PROBE_INTERVAL_MS);
+            measurement.nextProbeDelayMillis = 0;
+        }
+
+        if (mbeanRegistered)
+            MBeanWrapper.instance.unregisterMBean(mbeanName);
+
+        mbeanRegistered = false;
+    }
+
+    private static void cancelAndWait(ScheduledFuture future, long timeout, TimeUnit unit)
+    {
+        if (future != null)
+        {
+            future.cancel(false);
+            try
             {
-                // we do this so that a host considered bad has a chance to recover, otherwise would we never try
-                // to read from it, which would cause its score to never change
-                reset();
+                future.get(timeout, unit);
             }
-        };
+            catch (CancellationException | InterruptedException | ExecutionException | TimeoutException ignored)
+            {
+                // Exception is expected to happen eventually due to the cancel -> get
```

Comment:
I believe you are correct, but we also don't care if the latency probes are interrupted or throw and exception (we're just trying to tear down the futures). I'll update the comment accordingly.

Reply:
So I just realized this exposes the lack of exception handling for these scheduled tasks. They each need top level exception handling. If they don't have it they will stop without warning and the error will be swallowed up by the scheduled future until something calls get(). They should probably have a top level try catch that logs the error and doesn't let it escape so it will continue to retry on schedule.

That also means that all these other exception types again become an unexpected error except for CancellationException. I think it's fine to continue if you get an error, but we should log it.